### PR TITLE
feat: support files in drive +add-comment

### PR DESCRIPTION
**Pull request by wittam-01**
## Summary
Extend Drive file support in `drive +add-comment`. File targets are limited to full comments on supported Drive file extensions, with metadata validation before calling the comment API.

## Changes
- Support Drive file URLs and bare file tokens via `--type file` in `drive +add-comment`.
- Maintain a supported file-extension whitelist for file comments: `.md`, `.txt`, `.json`, `.csv`, `.go`, `.js`, `.py`, `.pptx`, `.png`, `.jpg`, `.zip`, `.mp3`, `.mp4`.
- Add a `metas.batch_query` preflight to read the target file title and verify its extension is supported.
- Reject unsupported file types, including `.pdf`, `.docx`, and `.xlsx`, before calling the comment API because those page previews currently display comments incorrectly.
- Reject file comments that try to use `--block-id` or `--selection-with-ellipsis`, since Drive file targets only support full comments.
- Create Drive file comments with `file_type=file` and a fixed placeholder `anchor.block_id` value of `test`.
- Return the validated `file_extension` in the shortcut output.
- Update `lark-drive` skill documentation with file comment examples, supported-extension constraints, and the full-comment-only limitation.
- Add unit coverage, dry-run E2E coverage, and a gated live E2E workflow for file comments.

## Test Plan
- [x] `make build`
- [x] `go test ./shortcuts/drive -count=1`
- [x] `go test ./tests/cli_e2e/drive -run TestDriveAddCommentDryRun_File -count=1`

## Related Issues
- None

## Summary by CodeRabbit

* **New Features**
  * Drive files now support full-document comments (in addition to docs/sheets/slides); only full comments are allowed for file targets.

* **Behavior Changes**
  * CLI rejects local/selection comment flags for file targets and enforces a whitelist of supported file extensions.
  * Dry-run now previews the metadata check and the file comment creation flow.

* **Documentation**
  * Updated command docs and examples for commenting on Drive files.

* **Tests**
  * Added end-to-end and unit tests covering validation, execution, dry-run, and supported-extension checks.

[Review Change Stack](https://app.coderabbit.ai/change-stack/larksuite/cli/pull/975?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: shortcuts/drive/drive_add_comment.go [hunk 1]
```diff
@@ -47,6 +47,31 @@ const defaultLocateDocLimit = 10
...
+	".txt",
+	".json",
+	".csv",
+	".go",
+	".js",
+	".py",
+	".pptx",
+	".png",
+	".jpg",
+	".zip",
```
Comment (fangshuyu-768) [inline shortcuts/drive/drive_add_comment.go:71]: **P1: Missing `.jpeg` extension in the whitelist**

The whitelist includes `.jpg` but omits `.jpeg`. macOS and many cameras produce JPEG files with the `.jpeg` extension, which would be incorrectly rejected.

```suggestion
	".jpg",
	".jpeg",
	".zip",
```

File: shortcuts/drive/drive_add_comment.go [hunk 2]
```diff
@@ -809,6 +896,84 @@ func parseSheetCellRef(input string) (*sheetAnchor, error) {
...
+	)
+	if err != nil {
+		return "", err
+	}
+
+	metas := common.GetSlice(data, "metas")
+	if len(metas) == 0 {
+		return "", output.Errorf(output.ExitAPI, "api_error", "drive metas.batch_query returned no metadata for file %s", common.MaskToken(fileToken))
+	}
+	meta, _ := metas[0].(map[string]interface{})
```
Comment (fangshuyu-768) [inline shortcuts/drive/drive_add_comment.go]: **P1: Silent type assertion failure**

The `ok` value is discarded. If `metas[0]` is not `map[string]interface{}`, `meta` is nil, `common.GetString(nil, "title")` returns `""`, and the user gets a misleading "file metadata did not return a title" error instead of a more precise type-mismatch message.

```suggestion
	meta, ok := metas[0].(map[string]interface{})
	if !ok {
		return "", output.Errorf(output.ExitAPI, "api_error", "drive metas.batch_query returned unexpected metadata format for file %s", common.MaskToken(fileToken))
	}
	return common.GetString(meta, "title"), nil
```

**Review by fangshuyu-768 — COMMENTED** [review]
## PR #975 Review: `feat: support files in drive +add-comment`

Overall the implementation is clean and well-tested. The two-step orchestration (metadata verify → create comment) is a good design, and test coverage is solid. Found 2 P1 issues that should be fixed before merge, plus a few lower-priority items.

### P1 (should fix before merge)
1. **Missing `.jpeg` extension** in the whitelist — `.jpg` is present but `.jpeg` is not, which will incorrectly reject legitimate JPEG files.
2. **Silent type assertion failure** in `fetchCommentTargetFileTitle` — discarding `ok` from the type assertion produces a misleading error message when the assertion fails.

### P2 (nice to fix)
3. Duplicated "file only supports full comments" validation in 3 places — extract a helper.
4. `fileCommentExtension` returns unexpected result for dotfiles like `.gitignore`.
5. Placeholder `"test"` lacks a TODO/issue tracking reference.

### P3 (optional)
6. Extension lookup uses linear scan — a map would be more idiomatic.
7. Redundant `strings.ToLower` in `isSupportedFileCommentExtension`.

File: shortcuts/drive/drive_add_comment.go [hunk 3]
```diff
@@ -47,6 +47,31 @@ const defaultLocateDocLimit = 10
 // with `drive file.comments create_v2` against a fresh docx.
 const maxCommentTotalRunes = 10000
 
+// The file comment API treats supported Drive file comments as full-file
+// comments in the UI, but currently rejects an empty anchor.block_id for file
+// targets. Use the backend-accepted placeholder until the API supports
+// omitting it.
+const fileFullCommentAnchorBlockID = "test"
```
Comment (fangshuyu-768) [inline shortcuts/drive/drive_add_comment.go:54]: **P2: Placeholder lacks a tracking reference**

The comment explains this is a workaround, but there's no TODO or issue reference. Once the backend API supports an empty `block_id`, this placeholder should be removed — easy to forget without a tracking marker.

Consider adding a `TODO(issue#xxx):` or at least `TODO:` in the comment.

File: shortcuts/drive/drive_add_comment.go [hunk 4]
```diff
@@ -451,9 +516,14 @@ func resolveCommentTarget(ctx context.Context, runtime *common.RuntimeContext, i
...
 
-	if docRef.Kind == "docx" || docRef.Kind == "doc" || docRef.Kind == "sheet" || docRef.Kind == "slides" {
-		if mode == commentModeLocal && docRef.Kind != "docx" && docRef.Kind != "sheet" && docRef.Kind != "slides" {
-			return resolvedCommentTarget{}, output.ErrValidation("local comments only support docx, sheet, and slides; old doc format only supports full comments")
+	if docRef.Kind == "docx" || docRef.Kind == "doc" || docRef.Kind == "file" || docRef.Kind == "sheet" || docRef.Kind == "slides" {
+		if mode == commentModeLocal {
+			switch docRef.Kind {
+			case "doc":
+				return resolvedCommentTarget{}, output.ErrValidation("local comments only support docx, sheet, and slides; old doc format only supports full comments")
+			case "file":
```
Comment (fangshuyu-768) [inline shortcuts/drive/drive_add_comment.go:523]: **P2: Duplicated "file only supports full comments" validation**

The same check appears in 3 places:
1. Validate stage (~L175): `if docRef.Kind == "file"` checks `--block-id` and `--selection-with-ellipsis`
2. Here: `case "file"` returns the same error
3. Wiki-to-file path (~L585): `if objType == "file"` returns the same error

If the constraint changes, all three must be updated in sync — easy to miss one. Consider extracting a helper like `validateFileCommentMode(mode string) error`.

File: shortcuts/drive/drive_add_comment.go [hunk 5]
```diff
@@ -809,6 +896,84 @@ func parseSheetCellRef(input string) (*sheetAnchor, error) {
...
+		fmt.Sprintf("drive +add-comment does not support comments for this Drive file type yet; got %q (%s)", title, extensionLabel),
+		"file comments currently support full comments only for these extensions: "+supportedFileCommentExtensionsText(),
+	)
+}
+
+func fileCommentExtension(title string) string {
+	title = strings.TrimSpace(title)
+	idx := strings.LastIndex(title, ".")
+	if idx < 0 || idx == len(title)-1 {
+		return ""
```
Comment (fangshuyu-768) [inline shortcuts/drive/drive_add_comment.go:969]: **P2: `fileCommentExtension` returns unexpected result for dotfiles**

For a dotfile like `.gitignore`, `strings.LastIndex` returns 0, so the function returns `.gitignore` as the "extension". While this won't match the whitelist (no false pass), the error message would display a confusing extension label instead of "no extension".

Consider treating `idx == 0` as no extension:

```suggestion
	if idx <= 0 || idx == len(title)-1 {
```

File: shortcuts/drive/drive_add_comment.go [hunk 6]
```diff
@@ -809,6 +896,84 @@ func parseSheetCellRef(input string) (*sheetAnchor, error) {
...
+	}
+	return strings.ToLower(title[idx:])
+}
+
+func isSupportedFileCommentExtension(extension string) bool {
+	extension = strings.ToLower(strings.TrimSpace(extension))
```
Comment (fangshuyu-768) [inline shortcuts/drive/drive_add_comment.go]: **P3: Redundant `strings.ToLower` + linear scan**

`fileCommentExtension` already lowercases the extension. The second `strings.ToLower` here is redundant. Also, for a 13-element list a `map[string]struct{}` initialized once would be more idiomatic Go and O(1) lookup.

File: shortcuts/drive/drive_add_comment.go (hunk 1 above)
Comment (wittam-01) [reply shortcuts/drive/drive_add_comment.go:71]: Fixed in 787bff9d: added `.jpeg` to the supported file-comment extension whitelist and synced the skill/reference docs. Covered by `TestFileCommentSupportedExtensions` and verified live with bot identity on a `.jpeg` file (`comment_id=7642317576034372551`).

File: shortcuts/drive/drive_add_comment.go (hunk 3 above)
Comment (wittam-01) [reply shortcuts/drive/drive_add_comment.go:54]: Fixed in 787bff9d: added a TODO on the file full-comment placeholder explaining it should be removed once the API accepts omitting `anchor.block_id` for file full comments.

File: shortcuts/drive/drive_add_comment.go (hunk 2 above)
Comment (wittam-01) [reply shortcuts/drive/drive_add_comment.go]: Fixed in 787bff9d: `fetchCommentTargetFileTitle` now checks the `metas[0]` type assertion and returns an explicit unexpected metadata format error. Added `TestFileCommentExecuteRejectsUnexpectedMetadataFormat`.

File: shortcuts/drive/drive_add_comment.go (hunk 4 above)
Comment (wittam-01) [reply shortcuts/drive/drive_add_comment.go:523]: Addressed in 787bff9d: extracted `validateFileCommentMode` and reused it for direct file targets and wiki-to-file targets so the full-comment-only constraint has a single implementation point.

File: shortcuts/drive/drive_add_comment.go (hunk 5 above)
Comment (wittam-01) [reply shortcuts/drive/drive_add_comment.go:969]: Fixed in 787bff9d: `fileCommentExtension` now treats dotfiles (`idx == 0`) as having no extension. Added `.gitignore` coverage in `TestFileCommentSupportedExtensions`.

File: shortcuts/drive/drive_add_comment.go (hunk 6 above)
Comment (wittam-01) [reply shortcuts/drive/drive_add_comment.go]: Addressed in 787bff9d: added `supportedFileCommentExtensionSet` for map lookup and removed the redundant lowercasing from `isSupportedFileCommentExtension`; normalization stays in `fileCommentExtension`.

**Comment by wittam-01** [comment]
Follow-up fixes are amended into 787bff9d.

Addressed review items:
- Added `.jpeg` to the Drive file comment whitelist and docs.
- Added explicit `metas[0]` type assertion handling for unexpected metadata format.
- Extracted `validateFileCommentMode` for file full-comment-only validation.
- Treat dotfiles like `.gitignore` as no-extension.
- Added TODO for the temporary file `anchor.block_id` placeholder.
- Switched supported-extension lookup to a map.
- Also updated the unsupported wiki-target error to include `file`.

Validation:
- `go test ./shortcuts/drive -count=1`
- `make build`
- `go test ./tests/cli_e2e/drive -run TestDriveAddCommentDryRun_File -count=1`
- Live bot verification in the Drive test folder: `.jpeg`, `.txt`, `.json`, `.md` file comments all succeeded.

**Comment by wittam-01** [comment]
Follow-up edge case fix amended into 7dd2916c: a file whose full title is exactly a whitelisted extension such as `.md` is now treated as that extension, while unsupported dotfiles such as `.gitignore` still resolve to no extension.

Additional validation:
- Uploaded a real file named `.md` with bot identity.
- Before the fix it was rejected as `got ".md" (no extension)`.
- After the fix, `drive +add-comment` succeeded with `file_extension=.md`, `comment_id=7642320093141093336`.
- `go test ./shortcuts/drive -count=1`
- `go test ./tests/cli_e2e/drive -run TestDriveAddCommentDryRun_File -count=1`

**Review by fangshuyu-768 — APPROVED** [review]
